charon/src/ast/types_utils.rs

```diff
@@ -901,6 +913,66 @@ pub trait TyVisitable: Sized + AstVisitable {
     }
 }
 
+impl TypeDecl {
+    /// Computes the variant from the tag.
+    ///
+    /// If the `tag` does not correspond to any valid discriminant but there is a niche,
+    /// the resulting `VariantId` will be for the untagged variant[`TagEncoding::Niche::untagged_variant`].
+    pub fn get_variant_from_tag(&self, tag: ScalarValue) -> Option<VariantId> {
+        let layout = self.layout.as_ref()?;
+        if layout.uninhabited {
+            return None;
+        };
+        let discr_layout = layout.discriminant_layout.as_ref()?;
+        match &self.kind {
+            TypeDeclKind::Enum(variants) => {
+                if variants.is_empty() {
+                    None
+                } else {
+                    let discr_ty = variants.get(VariantId::ZERO)?.discriminant.get_integer_ty();
+
+                    match &discr_layout.encoding {
+                        TagEncoding::Direct => {
+                            assert_eq!(tag.get_integer_ty(), discr_layout.tag_ty);
+                            let discr = ScalarValue::from_bits(discr_ty, tag.to_bits());
+                            // Find corresponding variant.
+                            variants.iter_indexed().find_map(|(id, variant)| {
```

Comment:
We must use `variant_layouts` in both cases. The fact that discriminant values equal tag values is an implementation detail I don't want to rely on.